### PR TITLE
Demonstrates duckdb adapters + data loaders

### PR DESCRIPTION
**Pull request by elijahbenizzy**
[Short description explaining the high-level reason for the pull request]

## Changes

- Adds documentation on duckdb integration
- Adds documentation on data-loaders as adaptable modules

## Testing

1.

## Checklist

- [ ] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code can be automatically merged (no conflicts)
- [ ] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [ ] Passes all existing automated tests
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future TODOs are captured in comments
- [ ] Project documentation has been updated if adding/changing functionality.
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Python - local testing

- [ ] python 3.6
- [ ] python 3.7

### CODE REVIEW

File: examples/data_loaders/load_data_duckdb.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import duckdb
+import pandas as pd
+
+
+def connection(db_path: str) -> duckdb.DuckDBPyConnection:
+    return duckdb.connect(database=db_path)
+
+
+def spend(connection: duckdb.DuckDBPyConnection) -> pd.DataFrame:
+    return connection.execute("select * from marketing_spend").fetchdf()
```
Comment (skrawcz) [inline examples/data_loaders/load_data_duckdb.py:10]: cool -- any plans to implement an idea from your doc about reducing this boilerplate?

File: examples/data_loaders/generate_test_data.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    }
+    return pd.DataFrame(data)
+
+
+@click.group()
+def main():
+    pass
+
+
+class MaterializationDriver(driver.Driver, abc.ABC):
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply examples/data_loaders/generate_test_data.py:121]: oh I see that this isn't used and was likely your experimenting?

File: examples/data_loaders/README.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
+# Data Loaders
+
+Among multiple uses, Hamilton excels at building maintainble, scalable reprsentations of ETLs.
+If you've read through the other guides, it should be pretty clear how hamilton enables transformations (the T in ETL).
```
Comment (skrawcz) [inline examples/data_loaders/README.md:4]:
```suggestion
If you've read through the other guides, it should be pretty clear how hamilton enables transformations (the T in ETL/ELT).
```

File: examples/data_loaders/README.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+In this example, we'll talk about an approach to _Extracting_ data, and how Hamilton enables you to build out extracts,
+in a scalable, pluggable way.
+
+Note that while we're still coming up with higher-level abstractions, the approach outlined here is recommended.
```
Comment (skrawcz) [inline examples/data_loaders/README.md:8]: Not needed?
```suggestion
```

File: examples/data_loaders/README.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
+# Data Loaders
+
+Among multiple uses, Hamilton excels at building maintainble, scalable reprsentations of ETLs.
+If you've read through the other guides, it should be pretty clear how hamilton enables transformations (the T in ETL).
+In this example, we'll talk about an approach to _Extracting_ data, and how Hamilton enables you to build out extracts,
+in a scalable, pluggable way.
```
Comment (skrawcz) [inline examples/data_loaders/README.md:6]: need to add more on the value of doing things this way... something like:
```suggestion
in a scalable, pluggable way. For example, being able to switch where data is loaded between development and production is useful, since you might only want a subsample in development, or even loading it from a different source. Here we'll show you how you can achieve this without cluttering your code with `if else`, which will make your dataflow easier to maintain in the long run as well.
```

File: examples/data_loaders/generate_test_data.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+
+class DuckDBMaterializationDriver(MaterializationDriver):
+    def __init__(self, path: str, config: Dict[str, Any], modules: List[ModuleType]):
+        super(DuckDBMaterializationDriver, self).__init__(config, *modules)
+        self.con = duckdb.connect(database=path, read_only=False)
+
+    def materialize(self, df: pd.DataFrame, table: str):
+        self.con.execute(f"CREATE TABLE {table} AS SELECT * from df")
+        self.con.fetchall()
```
Comment (skrawcz) [inline examples/data_loaders/generate_test_data.py:176]: no return statement?

File: examples/data_loaders/generate_test_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (skrawcz) [reply examples/data_loaders/generate_test_data.py:122]: Yeah hmm... need to think carefully about this. People cut and paste code...

Otherwise `materialize` here should mean saving -- not executing a query?

File: examples/data_loaders/generate_test_data.py (hunk 6 above)
Comment (elijahbenizzy) [reply examples/data_loaders/generate_test_data.py:176]: Nope, it saves it.

File: examples/data_loaders/generate_test_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply examples/data_loaders/generate_test_data.py:122]: Yeah, so I'm happy to do this another way. This was purely for my own internal iteration (E.G. creating the data) -- its not really meant to be part of an example, but to your point people will probably just use it anyway.

That said, `materialize` does mean saving -- saving is done by executing a query.

File: examples/data_loaders/generate_test_data.py (hunk 6 above)
Comment (skrawcz) [reply examples/data_loaders/generate_test_data.py:176]: Hmm -- yeah I misread this.

I have being using `materialize` to mean it gets created and put into memory. I think saving it should instead use language that is less ambiguous:

* export
* save
* load (like ETL)
* materialize_to_store

File: examples/data_loaders/README.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+In this example, we'll talk about an approach to _Extracting_ data, and how Hamilton enables you to build out extracts,
+in a scalable, pluggable way. For example, being able to switch where data is loaded between development and production
+is useful, since you might only want a subsample in development, or even load it from a different source.
+Here we'll show you how you can achieve this without cluttering your code with `if else`,
+which will make your dataflow easier to maintain in the long run.
+
+The goal is to show you two things:
+
+1. How to load data from various sources
+2. How to switch the sources of data you're loading from by swapping out modules, using[polymorphism](https://en.wikipedia.org/wiki/Polymorphism_(computer_science))
```
Comment (skrawcz) [inline examples/data_loaders/README.md:14]: I'm confused with the "saving" back out to them Driver that you've added here? The "MaterializationDriver" seems orthogonal here?

File: examples/data_loaders/run.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    vars = driver.list_available_variables()
+    print(driver.execute([var.name for var in vars]))
+
+
+@main.command()
+def mock():
+    driver = hamilton.driver.Driver({}, load_data_mock, prep_data)
+    vars = driver.list_available_variables()
+    print(
+        driver.raw_execute(
```
Comment (skrawcz) [inline examples/data_loaders/run.py:37]: why `raw_execute`?

File: examples/data_loaders/generate_test_data.py (hunk 6 above)
Comment (elijahbenizzy) [reply examples/data_loaders/generate_test_data.py]: Removing it anyway, but sure, that language works

**Review by skrawcz — COMMENTED** [review]
also should remove the database.duckdb file?

**Comment by elijahbenizzy** [comment]
> also should remove the database.duckdb file?

Nope we deleted the generator so it has to be there. We load from it.

**Review by skrawcz — APPROVED** [review]
